### ISSUE TITLE
Node drain gets stuck when first OSD drained has no PGs assigned to it and using PDBs

### ISSUE THREAD

**Issue by jrcichra** [rook/rook#12375]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
When performing a drain operation on a Rook cluster with a large # of disks but little to no data and managed pod disruption budgets are enabled, drain operations will hang forever.

This is an edge case that's easy to work around (add enough `/dev/zero` data to the cluster so every OSD has at least one PG assigned to it).

**Expected behavior:**
A large, freshly made, no space used cluster can use managed pod disruption budgets and safely be drained.

**How to reproduce it (minimal and precise):**

Make a cluster (should be doable in VMS too) with a fairly large number of nodes and OSDs and do not write any data to it.
In my environment I had 15 nodes with 10 NVME disks each, 2 OSDs per NVME totaling 300 OSDs.

Make sure the managed pod disruption budgets are on.

Issue a kubectl drain without force so the PDBs will be respected.

One OSD will be picked to drain and draining that OSD will have no affect on the health of the PGs. The drain does not detect a node drain and the drain is stuck.

**File(s) to submit**:

* Cluster CR (custom resource), typically called `cluster.yaml`, if necessary
```
apiVersion: ceph.rook.io/v1
kind: CephCluster
metadata:
  name: rook-ceph
  namespace: rook-ceph-nvme
spec:
  cephVersion:
    allowUnsupported: false
    image: <REDACTED>
  cleanupPolicy:
    allowUninstallWithVolumes: false
    sanitizeDisks:
      dataSource: zero
      iteration: 1
      method: quick
  continueUpgradeAfterChecksEvenIfNotHealthy: false
  crashCollector:
    disable: true
  dashboard:
    enabled: true
    ssl: false
  dataDirHostPath: /state/var/lib/rook
  disruptionManagement:
    manageMachineDisruptionBudgets: false
    managePodBudgets: true
    osdMaintenanceTimeout: 50
  healthCheck:
    daemonHealth:
      mon:
        disabled: false
        interval: 45s
      osd:
        disabled: false
        interval: 60s
      status:
        disabled: false
        interval: 60s
    livenessProbe:
      mgr:
        disabled: false
      mon:
        disabled: false
      osd:
        disabled: false
  mgr:
    modules:
      - enabled: true
        name: pg_autoscaler
  mon:
    allowMultiplePerNode: false
    count: 5
  monitoring:
    enabled: false
  network:
    connections:
      encryption:
        enabled: true
    hostNetwork: true
  removeOSDsIfOutAndSafeToRemove: false
  resources:
    osd:
      limits:
        cpu: "2"
        memory: 8Gi
      requests:
        cpu: "2"
        memory: 8Gi
  skipUpgradeChecks: false
  storage:
    config:
      deviceClass: nvme
      encryptedDevice: "true"
      osdsPerDevice: "2"
    deviceFilter: ^nvme
    useAllNodes: true

```
**Logs to submit**:

* Operator's logs, if necessary
This is the main message we see:
```
clusterdisruption-controller: osd is down in the failure domain "C109", but pgs are active+clean. Requeuing in case pg status is not updated yet...
```
My hunch is I'm hitting both ends of this if block:
https://github.com/rook/rook/blob/8ea26a72ed1be40afc12a196704f4c6537754ea4/pkg/operator/ceph/disruption/clusterdisruption/osd.go#L351 and never progressing because the OSD which drained does not have any PGs assigned to it.

* Crashing pod(s) logs, if necessary

  To get logs, use `kubectl -n <namespace> logs <pod name>`
  When pasting logs, always surround them with backticks or use the `insert code` button from the Github UI.
  Read [GitHub documentation if you need help](https://help.github.com/en/articles/creating-and-highlighting-code-blocks).

**Cluster Status to submit**:
* Output of krew commands, if necessary

  To get the health of the cluster, use `kubectl rook-ceph health`
  To get the status of the cluster, use `kubectl rook-ceph ceph status`
  For more details, see the [Rook Krew Plugin](https://rook.io/docs/rook/latest-release/Troubleshooting/krew-plugin)

**Environment**:
* OS (e.g. from /etc/os-release): Debian Buster
* Kernel (e.g. `uname -a`): Variant of 6.1.27
* Cloud provider or hardware configuration: Bare Metal
* Rook version (use `rook version` inside of a Rook Pod): v1.9.2
* Storage backend version (e.g. for ceph do `ceph -v`): ceph version 16.2.7
* Kubernetes version (use `kubectl version`): v1.23.15
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): K8S from scratch
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/latest-release/Troubleshooting/ceph-toolbox/#interactive-toolbox)): HEALTH_OK

This is a low priority issue for me, but wanted to bring up the edge case in case there's a good way for us to address it. I understand there are limitations with the disruption detection system.

Perhaps after some period of time, even with all healthy PGs, we proceed with the PDBs set to allow only that node to fully drain?

**Comment by travisn** [comment]
@sp98 In the following code block, do you recall why we have different behavior if the PGs are clean? Seems like whether or not the PGs are clean, we could allow the drain to proceed, instead of resetting the PGs and blocking the drain. 
https://github.com/rook/rook/blob/release-1.12/pkg/operator/ceph/disruption/clusterdisruption/osd.go#L341-L353

**Comment by davidjsherman** [comment]
Could anyone suggest a workaround, other than manually filling the OSDs to assign PGs?

**Comment by travisn** [comment]
@davidjsherman As a workaround, you could disable the PDBs with this setting in the cluster CR:
```yaml
  disruptionManagement:
    managePodBudgets: false
```

@sp98 Shall we allow draining to proceed in this case?

**Comment by sp98** [comment]
@davidjsherman @travisn 
Rook reads the pg status from `ceph status` output.
When an OSD is drained (or goes down due to some other reasons), the PG status might be updated instantaneously in the `ceph status` output. 

[Here](https://github.com/rook/rook/blob/3dd76e061e8e96ed3f5214ccffe33a7116e5cd9e/pkg/operator/ceph/disruption/clusterdisruption/osd.go#L341-L353), rook waits for 30 seconds after the OSD drain to verify if the PGs have gone bad. 

For example: If an OSD was drained at, say, `12:00:00`, but the PGs are still active+clean till `12:00:30`, then rook will consider cluster to be healthy and allow more OSDs to drain.

**Comment by sp98** [comment]
@davidjsherman I didn't get a chance to look into it sooner. Any chance you can still share more logs for this behavior. I'm looking for following data after you start the node drain and it gets stuck:

- rook-ceph-operator logs with debug mode on in a text file.
-`rook-ceph-pdbstatemap` config map

**Comment by Ahmed-Kallel** [comment]
@sp98 @travisn 
I'm actively collaborating with @davidjsherman on this. You can find the requested files at the following link: 
https://filesender.renater.fr/?s=download&token=e1c7e13d-8e12-4c5d-81c9-66940cac6c1b

Thank you

**Comment by travisn** [comment]
This would be fixed with #13511.

**Comment by davidjsherman** [comment]
@travisn thanks for the heads up, @Ahmed-Kallel and I will check it out!
